core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -134,8 +134,11 @@ private[spark] class Executor(
     env.metricsSystem.registerSource(new JVMCPUSource())
     executorMetricsSource.foreach(_.register(env.metricsSystem))
     env.metricsSystem.registerSource(env.blockManager.shuffleMetricsSource)
+  } else {
+    Executor.executorSource = executorSource
```

Comment:
What happens when we call `env.metricsSystem.registerSource(executorSource)` here?

Reply:
Good question, this is the actual pain point this PR tries to solve: one cannot simply register metrics on env here when running in local mode, or otherwise the appId would not be available, so the resulting output would not be clearly usable (missing a key piece of info as the appId). That's why I propose to register the metrics in the Spark Context, together with other driver metrics. As you can see other metrics namespaces use a similar strategy.
To get around the issue, for local mode, I propose using the workaround of storing the executorSource in the object so it can be read later, I hope this is acceptable.

Reply:
I can think of other ways to do this more cleanly but it would be quite a bit more code change.  I do think we should explicitly make a comment about the appId so that someone who comes and looks at this later realizes that